chain/indexer/src/streamer/streamer.rs

```diff
@@ -84,39 +92,64 @@ async fn fetch_block_by_height(
         .map_err(|err| FailedToFetchData::String(err))
 }
 
-/// This function supposed to return the entire `BlockResponse`.
-/// It fetches the block and fetches all the chunks for the block
+/// This function supposed to return the entire `StreamerMessage`.
+/// It fetches the block and all related parts (chunks, outcomes, state changes etc.)
 /// and returns everything together in one struct
-async fn fetch_block_response(
+async fn build_streamer_message(
     client: &Addr<near_client::ViewClientActor>,
     block: views::BlockView,
-    outcomes_to_get: impl Iterator<Item = types::TransactionOrReceiptId>,
-) -> Result<FetchBlockResponse, FailedToFetchData> {
+) -> Result<StreamerMessage, FailedToFetchData> {
     let chunks = fetch_chunks(&client, &block.chunks).await?;
-    let (outcomes, mut outcomes_to_retry) = fetch_outcomes(&client, outcomes_to_get).await;
 
-    for chunk in &chunks {
-        outcomes_to_retry.extend(chunk.transactions.iter().map(|transaction| {
-            types::TransactionOrReceiptId::Transaction {
-                transaction_hash: transaction.hash.clone(),
-                sender_id: transaction.signer_id.to_string(),
-            }
-        }));
+    let mut local_receipts: Vec<views::ReceiptView> = vec![];
+    let mut outcomes = fetch_outcomes(&client, block.header.hash).await?;
+    let mut indexer_chunks: Vec<IndexerChunkView> = vec![];
 
-        outcomes_to_retry.extend(chunk.receipts.iter().map(|receipt| {
-            types::TransactionOrReceiptId::Receipt {
-                receipt_id: receipt.receipt_id,
-                receiver_id: receipt.receiver_id.to_string(),
-            }
-        }));
+    for chunk in &chunks {
+        let indexer_transactions = chunk
+            .transactions
+            .iter()
+            .map(|transaction| {
+                if transaction.signer_id == transaction.receiver_id {
```

Comment:
I suggest having a separate iteration over the list with `.filter(|transaction| transaction.signer_id == transaction.receiver_id).map()` and collect into the `local_receipt` list instead of abusing `map` with internal `pushing to an external vector of `local_receipts`. It will be easier to reason about.